substrate/frame/support/procedural/src/pallet/expand/event.rs

```diff
@@ -96,6 +95,17 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
 		event_item.variants.push(variant);
 	}
 
+	let deprecation = crate::deprecation::get_deprecation_enum(
+		&quote::quote! {#frame_support},
+		&event.attrs,
+		event_item
+			.variants
+			.iter()
+			.enumerate()
+			.map(|(index, item)| (index as u8, item.attrs.as_ref())),
```

Comment:
Same as error, some index can be given with `#[codec(index = N)]`

also interestingly such a code compiles, but I think it is not intended. (I don't think we need to support this. And macro should could return an error IMO)
```diff
diff --git a/substrate/frame/assets/src/lib.rs b/substrate/frame/assets/src/lib.rs
index 6dbce717a8e..f95d2ee2fc9 100644
--- a/substrate/frame/assets/src/lib.rs
+++ b/substrate/frame/assets/src/lib.rs
@@ -548,9 +548,10 @@ pub mod pallet {
 
        #[pallet::event]
        #[pallet::generate_deposit(pub(super) fn deposit_event)]
+       #[repr(u32)]
        pub enum Event<T: Config<I>, I: 'static = ()> {
                /// Some asset class was created.
-               Created { asset_id: T::AssetId, creator: T::AccountId, owner: T::AccountId },
+               Created { asset_id: T::AssetId, creator: T::AccountId, owner: T::AccountId } = 4,
                /// Some assets were issued.
                Issued { asset_id: T::AssetId, owner: T::AccountId, amount: T::Balance },
                /// Some assets were transferred.
```

Reply:
I think user of metadata only have access to information from scale-info. So if it is written that variant N is deprecated they will match it with the variant number N from the point of view of the scale-info information.

So yes enum variant changes position when sent of the wire. Because the wire doesn't transmit the rust variant index.

That said, I also don't mind forbidding usage of `#[codec(index = N)]` for both error and event variants

Reply:
Done, now when looking for `deprecated` attribute on `Event`'s or `Error`'s the code checks for discriminant or explicit `codec(index = N)` before using definition order.

Reply:
Yes, looks good to me